src/workerd/api/encoding.c++

```diff
@@ -460,38 +464,382 @@ kj::Maybe<jsg::JsString> TextDecoder::decodePtr(
+    auto length = str.length(js);
+    // Allocate buffer for Latin-1. Use v8::ArrayBuffer::NewBackingStore to avoid creating
+    // JS objects during conversion.
+    backingStore = js.allocBackingStore(length, jsg::Lock::AllocOption::UNINITIALIZED);
+    auto backingData = reinterpret_cast<kj::byte*>(backingStore->Data());
+
+    [[maybe_unused]] auto writeResult = str.writeInto(js, kj::arrayPtr(backingData, length));
+    KJ_DASSERT(
+        writeResult.written == length, "writeInto must completely overwrite the backing buffer");
+
+    utf8_length =
+        simdutf::utf8_length_from_latin1(reinterpret_cast<const char*>(backingData), length);
+
+    if (utf8_length == length) {
+      // ASCII fast path: no conversion needed, Latin-1 is same as UTF-8 for ASCII
+      auto array = v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore), 0, length);
+      return jsg::JsUint8Array(array);
+    }
+
+    KJ_DASSERT(utf8_length > length);
+
+    // Need to convert Latin-1 to UTF-8
+    std::shared_ptr<v8::BackingStore> backingStore2 =
+        js.allocBackingStore(utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
+    [[maybe_unused]] auto written =
+        simdutf::convert_latin1_to_utf8(reinterpret_cast<const char*>(backingData), length,
+            reinterpret_cast<char*>(backingStore2->Data()));
+    KJ_DASSERT(utf8_length == written);
+    auto array =
+        v8::Uint8Array::New(v8::ArrayBuffer::New(js.v8Isolate, backingStore2), 0, utf8_length);
+    return jsg::JsUint8Array(array);
+  }
+
+  // Two-byte string path
+  {
+    // Note that ValueView flattens the string, if it's not already flattened
+    v8::String::ValueView view(js.v8Isolate, str);
+    // Two-byte string path. V8 uses UTF-16LE encoding internally for strings with code points
+    // > U+00FF. Check if the UTF-16 is valid (no unpaired surrogates) to determine the path.
+    auto data = reinterpret_cast<const char16_t*>(view.data16());
+
+    if (simdutf::validate_utf16le(data, view.length())) {
+      // Common case: valid UTF-16, convert directly to UTF-8
+      utf8_length = simdutf::utf8_length_from_utf16le(data, view.length());
+      backingStore = js.allocBackingStore(utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
+      [[maybe_unused]] auto written = simdutf::convert_utf16le_to_utf8(
+          data, view.length(), reinterpret_cast<char*>(backingStore->Data()));
+      KJ_DASSERT(written == utf8_length);
+    } else {
+      // Invalid UTF-16 with unpaired surrogates. Per the Encoding Standard,
+      // unpaired surrogates must be replaced with U+FFFD (replacement character).
+      // Use custom conversion that handles invalid surrogates without creating an
+      // intermediate well-formed UTF-16 buffer.
+      auto inputArray = kj::ArrayPtr<const char16_t>(data, view.length());
+      utf8_length = utf8LengthFromInvalidUtf16(inputArray);
+      backingStore = js.allocBackingStore(utf8_length, jsg::Lock::AllocOption::UNINITIALIZED);
```

Comment:
hmm... question... what happens in the case allocating the backing store *does* trigger GC? Is the `ValueView` guaranteed to remain viable?

Reply:
Oops, yes that's bad.

I can't see why it wouldn't do a GC and that would ruin the ValueView.

It's a bit unexpected because creating the backing store doesn't allocate on the V8 heap, so you might not expect it to cause a GC, but if the external memory heuristics suggest it then it will GC :-/

Reply:
OK it would hit a release-mode assert because of the DisallowGarbageCollectionInRelease scope, so it's safe, but crashy.

Reply:
The best course of action is to be using kj::SmallArray rather than BackingStore, as James suggested. Implementing it now.

Reply:
hopefully this time, i've fixed this issue. thanks for the patience and the reviews.